### PR TITLE
[X86][AVX10.2] Support AVX10.2-BF16 new instructions. 

### PR DESCRIPTION
**Pull request by FreddyLeaf**
Ref.: https://cdrdv2.intel.com/v1/dl/getContent/828965

### CODE REVIEW

File: llvm/include/llvm/IR/IntrinsicsX86.td [hunk 1]
```diff
@@ -7219,3 +7219,413 @@ def int_x86_avx10_mask_vcvtneph2hf8s512 : ClangBuiltin<"__builtin_ia32_vcvtneph2
...
                               [IntrNoMem]>;
 }
+
+//===----------------------------------------------------------------------===//
+let TargetPrefix = "x86" in {
+  def int_x86_avx10_vaddnepbf16512
+      : ClangBuiltin<"__builtin_ia32_vaddnepbf16512">,
+        DefaultAttrsIntrinsic<[ llvm_v32bf16_ty ],
+                  [ llvm_v32bf16_ty, llvm_v32bf16_ty ],
+                  [ IntrNoMem ]>;
```
Comment (phoebewang) [inline llvm/include/llvm/IR/IntrinsicsX86.td:7229]: Indentation

File: clang/test/CodeGen/X86/avx10_2_512bf16-builtins.c [hunk 2]
```diff
@@ -0,0 +1,1054 @@
...
+
+__mmask32 test_mm512_cmp_pbh_mask_true_us(__m512bh a, __m512bh b) {
+  // CHECK-LABEL: test_mm512_cmp_pbh_mask_true_us
+  // CHECK: fcmp true <32 x bfloat> %{{.*}}, %{{.*}}
+  return _mm512_cmp_pbh_mask(a, b, _CMP_TRUE_US);
+}
+
+__mmask32 test_mm512_mask_cmp_pbh_mask_eq_oq(__mmask32 m, __m512bh a, __m512bh b) {
+  // CHECK-LABEL: @test_mm512_mask_cmp_pbh_mask_eq_oq
+  // CHECK: fcmp oeq <32 x bfloat> %{{.*}}, %{{.*}}
```
Comment (phoebewang) [inline clang/test/CodeGen/X86/avx10_2_512bf16-builtins.c:582]: Check select for mask version?

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 3]
```diff
@@ -14836,6 +14837,9 @@ Value *CodeGenFunction::EmitX86BuiltinExpr(unsigned BuiltinID,
   case X86::BI__builtin_ia32_vfmaddph512_mask:
   case X86::BI__builtin_ia32_vfmaddph512_maskz:
   case X86::BI__builtin_ia32_vfmaddph512_mask3:
+  case X86::BI__builtin_ia32_vfmaddnepbh128:
+  case X86::BI__builtin_ia32_vfmaddnepbh256:
+  case X86::BI__builtin_ia32_vfmaddnepbh512:
```
Comment (phoebewang) [inline clang/lib/CodeGen/CGBuiltin.cpp:14842]: Missing mask/maskz/mask3 variants.

File: clang/lib/Headers/avx10_2_512bf16intrin.h [hunk 4]
```diff
@@ -0,0 +1,565 @@
...
+}
+
+static __inline__ __m512bh __DEFAULT_FN_ATTRS512
+_mm512_maskz_rsqrt_pbh(__mmask32 __U, __m512bh __A) {
+  return (__m512bh)__builtin_ia32_vrsqrtpbf16512_mask(
+      (__v32bf)__A, (__v32bf)_mm512_setzero_pbh(), (__mmask32)__U);
+}
+
+#define _mm512_reducene_pbh(A, imm)                                            \
+  ((__m512bh)__builtin_ia32_vreducenepbf16512_mask(                            \
```
Comment (phoebewang) [inline clang/lib/Headers/avx10_2_512bf16intrin.h:381]: Missing Sema check fot the immediate.

File: clang/test/CodeGen/X86/avx10_2_512bf16-builtins.c (hunk 2 above)
Comment (FreddyLeaf) [reply clang/test/CodeGen/X86/avx10_2_512bf16-builtins.c:582]: There's only a following and. I can add a check for that. clang emit an extra and for _mask_cmp_ intrinsics, which is a gap with gcc: https://gcc.godbolt.org/z/5n63rYEqE

File: clang/lib/CodeGen/CGBuiltin.cpp (hunk 3 above)
Comment (FreddyLeaf) [reply clang/lib/CodeGen/CGBuiltin.cpp:14842]: this implementation didn't define them as bf16 fma instructions don't have rounding control, safe to implement by select.

File: llvm/lib/Target/X86/X86InstrFragmentsSIMD.td [hunk 5]
```diff
@@ -211,6 +211,12 @@ def X86CmpMaskCC :
       SDTypeProfile<1, 3, [SDTCisVec<0>, SDTCVecEltisVT<0, i1>,
                        SDTCisVec<1>, SDTCisSameAs<2, 1>,
                        SDTCisSameNumEltsAs<0, 1>, SDTCisVT<3, i8>]>;
+
+def X86CmpMaskCC_Int :
+      SDTypeProfile<1, 3, [SDTCisVec<0>, SDTCVecEltisVT<0, i1>,
+                       SDTCisVec<1>, SDTCisSameAs<2, 1>,
+                       SDTCisSameNumEltsAs<0, 1>, SDTCisVT<3, i32>]>;
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86InstrFragmentsSIMD.td:218]: Dead code, will remove

File: llvm/lib/Target/X86/X86InstrAVX10.td [hunk 6]
```diff
@@ -910,3 +910,313 @@ multiclass avx10_convert_2op_nomb<string OpcodeStr, AVX512VLVectorVTInfo _dest,
...
+                                T_MAP5, PD, EVEX_CD8<16, CD8VF>;
+    defm NEPBF16Z256 : avx512_fp_packed<opc, OpcodeStr, OpNode, MaskOpNode, 
+                                v16bf16x_info, sched.PH.YMM, IsCommutable>, EVEX_V256,
+                                T_MAP5, PD, EVEX_CD8<16, CD8VF>;
+  }
+}
+
+let Uses = []<Register>, mayRaiseFPException = 0 in {
+defm VADD : avx10_fp_binop_pbf16<0x58, "vaddne", fadd, SchedWriteFAddSizes, 1>;
+defm VSUB : avx10_fp_binop_pbf16<0x5C, "vsubne", fsub, SchedWriteFAddSizes, 0>;
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86InstrAVX10.td:962]: its parent multiclasses xmay set `mayRaiseFPException = 1`

File: clang/lib/Headers/avx10_2_512bf16intrin.h (hunk 4 above)
Comment (FreddyLeaf) [reply clang/lib/Headers/avx10_2_512bf16intrin.h:381]: [7957436](https://github.com/llvm/llvm-project/pull/101603/commits/7957436ac07d1902f3f510de756577c00353dd02)

File: clang/test/CodeGen/X86/avx10_2_512bf16-builtins.c (hunk 2 above)
Comment (FreddyLeaf) [reply clang/test/CodeGen/X86/avx10_2_512bf16-builtins.c:582]: [7957436](https://github.com/llvm/llvm-project/pull/101603/commits/7957436ac07d1902f3f510de756577c00353dd02)

File: llvm/include/llvm/IR/IntrinsicsX86.td (hunk 1 above)
Comment (FreddyLeaf) [reply llvm/include/llvm/IR/IntrinsicsX86.td]: [7957436](https://github.com/llvm/llvm-project/pull/101603/commits/7957436ac07d1902f3f510de756577c00353dd02)

File: llvm/lib/Target/X86/X86InstrAVX10.td [hunk 7]
```diff
@@ -910,3 +910,313 @@ multiclass avx10_convert_2op_nomb<string OpcodeStr, AVX512VLVectorVTInfo _dest,
...
+
+// VSCALEFPBF16
+multiclass avx10_fp_scalef_bf16<bits<8> opc, string OpcodeStr,
+                                X86SchedWriteWidths sched> {
+  let Predicates = [HasAVX10_2_512] in
+    defm PBF16Z : avx512_fp_scalef_p<opc, OpcodeStr, X86scalef, sched.ZMM, v32bf16_info>,
+                                EVEX_V512, T_MAP6,PS, EVEX_CD8<16, CD8VF>;
+  let Predicates = [HasAVX10_2] in {
+    defm PBF16Z128 : avx512_fp_scalef_p<opc, OpcodeStr, X86scalef, sched.XMM, v8bf16x_info>,
+                                   EVEX_V128, EVEX_CD8<16, CD8VF>, T_MAP6,PS;
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86InstrAVX10.td]: [7957436](https://github.com/llvm/llvm-project/pull/101603/commits/7957436ac07d1902f3f510de756577c00353dd02)

File: llvm/lib/Target/X86/X86InstrAVX10.td [hunk 8]
```diff
@@ -910,3 +910,313 @@ multiclass avx10_convert_2op_nomb<string OpcodeStr, AVX512VLVectorVTInfo _dest,
...
+                EVEX_B, Sched<[sched.Folded, sched.ReadAfterFold]>;
+  }
+}
+
+multiclass avx10_vcmp_bf16<X86SchedWriteWidths sched, AVX512VLVectorVTInfo _> {
+  let Predicates = [HasAVX10_2_512] in
+    defm Z   : avx10_vcmp_common_bf16<sched.ZMM, _.info512>, EVEX_V512;
+  let Predicates = [HasAVX10_2] in {
+   defm Z128 : avx10_vcmp_common_bf16<sched.XMM, _.info128>, EVEX_V128;
+   defm Z256 : avx10_vcmp_common_bf16<sched.YMM, _.info256>, EVEX_V256;
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86InstrAVX10.td]: [7957436](https://github.com/llvm/llvm-project/pull/101603/commits/7957436ac07d1902f3f510de756577c00353dd02)

File: llvm/lib/Target/X86/X86InstrAVX10.td [hunk 9]
```diff
@@ -910,3 +910,313 @@ multiclass avx10_convert_2op_nomb<string OpcodeStr, AVX512VLVectorVTInfo _dest,
...
                   AVX512XDIi8Base, T_MAP5, EVEX, EVEX_CD8<16, CD8VH>;
+
+//-------------------------------------------------
+// AVX10 BF16 instructions
+//-------------------------------------------------
+
+// VADDNEPBF16, VSUBNEPBF16, VMULNEPBF16, VDIVNEPBF16, VMAXPBF16, VMINPBF16
+multiclass avx10_fp_binopne_int_pbf16<bits<8> opc, string OpcodeStr,
+                   X86SchedWriteSizes sched,
+                   bit IsCommutable = 0> {
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86InstrAVX10.td]: [7957436](https://github.com/llvm/llvm-project/pull/101603/commits/7957436ac07d1902f3f510de756577c00353dd02)

File: llvm/lib/Target/X86/X86InstrAVX10.td [hunk 10]
```diff
@@ -910,3 +910,313 @@ multiclass avx10_convert_2op_nomb<string OpcodeStr, AVX512VLVectorVTInfo _dest,
...
+  let Predicates = [HasAVX10_2_512] in
+    defm PBF16Z : avx512_fp_packed<opc, OpcodeStr,
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86InstrAVX10.td]: [7957436](https://github.com/llvm/llvm-project/pull/101603/commits/7957436ac07d1902f3f510de756577c00353dd02)

File: llvm/lib/Target/X86/X86InstrAVX10.td [hunk 11]
```diff
@@ -910,3 +910,313 @@ multiclass avx10_convert_2op_nomb<string OpcodeStr, AVX512VLVectorVTInfo _dest,
...
+
+defm VCMPPBF16 : avx10_vcmp_bf16<SchedWriteFCmp, avx512vl_bf16_info>,
+                          AVX512XDIi8Base, EVEX, VVVV, EVEX_CD8<16, CD8VF>, TA;
+
+
+// VSQRTNEPBF16
+multiclass avx10_sqrt_packed_bf16<bits<8> opc, string OpcodeStr,
+                                  X86SchedWriteSizes sched> {
+  let Predicates = [HasAVX10_2_512] in
+  defm NEPBF16Z : avx512_sqrt_packed<opc, !strconcat(OpcodeStr, "pbf16"),
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86InstrAVX10.td]: [7957436](https://github.com/llvm/llvm-project/pull/101603/commits/7957436ac07d1902f3f510de756577c00353dd02)

File: llvm/lib/Target/X86/X86InstrAVX10.td [hunk 12]
```diff
@@ -910,3 +910,313 @@ multiclass avx10_convert_2op_nomb<string OpcodeStr, AVX512VLVectorVTInfo _dest,
...
+defm VMUL : avx10_fp_binop_pbf16<0x59, "vmulne", fmul, SchedWriteFMulSizes, 0>;
+defm VDIV : avx10_fp_binop_pbf16<0x5E, "vdivne", fdiv, SchedWriteFDivSizes, 0>;
+defm VMIN : avx10_fp_binopne_int_pbf16<0x5D, "vmin", SchedWriteFCmpSizes, 0>;
+defm VMAX : avx10_fp_binopne_int_pbf16<0x5F, "vmax", SchedWriteFCmpSizes, 0>;
+}
+
+// VCOMSBF16
+let Uses = []<Register>, mayRaiseFPException = 0,
+  Defs = [EFLAGS], Predicates = [HasAVX10_2_512] in {
+  defm VCOMSBF16Z : sse12_ord_cmp<0x2F, FR16X, null_frag, bf16, f16mem, loadf16,
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86InstrAVX10.td:974]: Added a todo [7957436](https://github.com/llvm/llvm-project/pull/101603/commits/7957436ac07d1902f3f510de756577c00353dd02)

File: llvm/lib/Target/X86/X86InstrAVX10.td [hunk 13]
```diff
@@ -910,3 +910,313 @@ multiclass avx10_convert_2op_nomb<string OpcodeStr, AVX512VLVectorVTInfo _dest,
...
+    defm NEPBF16Z256 : avx512_fp_packed<opc, OpcodeStr, OpNode, MaskOpNode, 
+                                v16bf16x_info, sched.PH.YMM, IsCommutable>, EVEX_V256,
+                                T_MAP5, PD, EVEX_CD8<16, CD8VF>;
+  }
+}
+
+let Uses = []<Register>, mayRaiseFPException = 0 in {
+defm VADD : avx10_fp_binop_pbf16<0x58, "vaddne", fadd, SchedWriteFAddSizes, 1>;
+defm VSUB : avx10_fp_binop_pbf16<0x5C, "vsubne", fsub, SchedWriteFAddSizes, 0>;
+defm VMUL : avx10_fp_binop_pbf16<0x59, "vmulne", fmul, SchedWriteFMulSizes, 0>;
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86InstrAVX10.td]: [7957436](https://github.com/llvm/llvm-project/pull/101603/commits/7957436ac07d1902f3f510de756577c00353dd02)

File: clang/lib/Headers/CMakeLists.txt [hunk 14]
```diff
@@ -147,11 +147,13 @@ set(x86_files
   amxcomplexintrin.h
   amxfp16intrin.h
   amxintrin.h
+  avx10_2_512bf16intrin.h
   avx10_2_512convertintrin.h
   avx10_2_512minmaxintrin.h
   avx10_2_512niintrin.h
   avx10_2_512satcvtintrin.h
   avx10_2convertintrin.h
+  avx10_2bf16intrin.h
```
[the remark this answers is not in the document]
Comment (FreddyLeaf) [reply clang/lib/Headers/CMakeLists.txt:155]: [7957436](https://github.com/llvm/llvm-project/pull/101603/commits/7957436ac07d1902f3f510de756577c00353dd02)

**Comment by phoebewang** [comment]
Should we add folding in X86InstrFMA3Info.cpp?

**Comment by FreddyLeaf** [comment]
> Should we add folding in X86InstrFMA3Info.cpp?

yes, addressed in [b9e35f0](https://github.com/llvm/llvm-project/pull/101603/commits/b9e35f050d6319ebd8da3c0f4cf5399dd03b3915)

File: llvm/lib/Target/X86/X86InstrFMA3Info.cpp [hunk 15]
```diff
@@ -153,7 +163,8 @@ const X86InstrFMA3Group *llvm::getFMA3Group(unsigned Opcode, uint64_t TSFlags) {
                         ((TSFlags & X86II::EncodingMask) == X86II::EVEX &&
                          ((TSFlags & X86II::OpMapMask) == X86II::T8 ||
                           (TSFlags & X86II::OpMapMask) == X86II::T_MAP6));
-  bool IsFMA3Prefix = (TSFlags & X86II::OpPrefixMask) == X86II::PD;
+  bool IsFMA3Prefix = ((TSFlags & X86II::OpPrefixMask) == X86II::PD) ||
+                      ((TSFlags & X86II::OpPrefixMask) == 0); // X86II::PS
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86InstrFMA3Info.cpp:167]: Remove the outer parentheses.

**Review by phoebewang — APPROVED** [review]
LGTM.

File: llvm/lib/Target/X86/X86InstrFMA3Info.cpp (hunk 15 above)
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86InstrFMA3Info.cpp]: [a6c4fce](https://github.com/llvm/llvm-project/pull/101603/commits/a6c4fce4069fb81ab4a50d825fe6039d45827d63)

File: llvm/lib/Target/X86/MCTargetDesc/X86ATTInstPrinter.cpp [hunk 16]
```diff
@@ -205,7 +214,8 @@ bool X86ATTInstPrinter::printVecCompareInstr(const MCInst *MI,
               printwordmem(MI, CurOp--, OS);
             else
               printdwordmem(MI, CurOp--, OS);
-          } else if ((Desc.TSFlags & X86II::OpPrefixMask) == X86II::XD) {
+          } else if ((Desc.TSFlags & X86II::OpPrefixMask) == X86II::XD &&
```
Comment (rupprecht) [inline llvm/lib/Target/X86/MCTargetDesc/X86ATTInstPrinter.cpp:217]: This check in X86IntelInstPrinter looks very similar, but isn't updated by this commit: https://github.com/llvm/llvm-project/blob/ede40da1f8c1e91601b985cd32ad785aa8806880/llvm/lib/Target/X86/MCTargetDesc/X86IntelInstPrinter.cpp#L207

We're getting an assertion there. Did you mean to also update the same check in X86IntelInstPrinter?

Comment (FreddyLeaf) [reply llvm/lib/Target/X86/MCTargetDesc/X86ATTInstPrinter.cpp:217]: Good catch. Will update. And the assert should be removed.

Comment (FreddyLeaf) [reply llvm/lib/Target/X86/MCTargetDesc/X86ATTInstPrinter.cpp:217]: addressed in https://github.com/llvm/llvm-project/pull/107529

**Pull request merged**